Packet++/src/TcpSorter.cpp

```diff
@@ -0,0 +1,585 @@
+	if(m_isClosed)
+	{
+		return;
+	}
+	// parse the packet
+	Packet packet(spRawPacket.get(), false);
+
+	time_t now = time(NULL);
+
+	// clean up any expired inactive TCP connection
+	if (0 != m_MaxIdleTimeout && m_LastCleanupTime + m_CleanUpInactiveConnPeriod < now)
+	{
+		cleanUpInactiveTcpConnection(now);
+		m_LastCleanupTime = now;
+	}
+
+	// get IP layer
+	Layer* ipLayer = nullptr;
+	if (packet.isPacketOfType(IPv4))
+		ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv4Layer>());
+	else if (packet.isPacketOfType(IPv6))
+		ipLayer = dynamic_cast<Layer*>(packet.getLayerOfType<IPv6Layer>());
+
+	if (ipLayer == nullptr)
+		return;
+
+	// Ignore non-TCP packets
+	TcpLayer* tcpLayer = packet.getLayerOfType<TcpLayer>();
+	if (tcpLayer == nullptr)
+		return;
+
+	// Ignore the packet if it's an ICMP packet that has a TCP layer
+	// Several ICMP messages (like "destination unreachable") have TCP data as part of the ICMP message.
+	// This is not real TCP data and packet can be ignored
+	if (packet.isPacketOfType(ICMP))
+	{
+		LOG_DEBUG("Packet is of type ICMP so TCP data is probably  part of the ICMP message. Ignoring this packet");
+		return;
+	}
+
+	// set the TCP payload size
+	uint32_t tcpPayloadSize =static_cast<uint32_t>(tcpLayer->getLayerPayloadSize());
+
+	// get TCP flag
+	bool isAck = (tcpLayer->getTcpHeader()->ackFlag == 1);
+	bool isSyn = (tcpLayer->getTcpHeader()->synFlag == 1);
+	bool isFin = (tcpLayer->getTcpHeader()->finFlag == 1);
+	bool isRst = (tcpLayer->getTcpHeader()->rstFlag == 1);
+
+	// TCP connection for captured packet
+	SPTcpSorterData tcpSorterData = nullptr;
+
+	// calculate flow key for this packet
+	uint32_t flowKey = hash5Tuple(&packet);
+
+	// find the connection in the connection map
+	ConnectionList::iterator iter = m_ConnectionList.find(flowKey);
+
+	// get packet's source and dest IP address
+	IPAddress* srcIP = nullptr;
+	IPAddress* dstIP = nullptr;
+	IPv4Address srcIP4Addr = IPv4Address::Zero;
+	IPv6Address srcIP6Addr = IPv6Address::Zero;
+	IPv4Address dstIP4Addr = IPv4Address::Zero;
+	IPv6Address dstIP6Addr = IPv6Address::Zero;
+	if (ipLayer->getProtocol() == IPv4)
+	{
+		srcIP4Addr = (dynamic_cast<IPv4Layer*>(ipLayer))->getSrcIpAddress();
+		srcIP = &srcIP4Addr;
+		dstIP4Addr = (dynamic_cast<IPv4Layer*>(ipLayer))->getDstIpAddress();
+		dstIP = &dstIP4Addr;
+	}
+	else if (ipLayer->getProtocol() == IPv6)
+	{
+		srcIP6Addr = (dynamic_cast<IPv6Layer*>(ipLayer))->getSrcIpAddress();
+		srcIP = &srcIP6Addr;
+		dstIP6Addr = (dynamic_cast<IPv6Layer*>(ipLayer))->getDstIpAddress();
+		dstIP = &dstIP6Addr;
+	}
+
+	// if TCP connection doesn't exist, create one
+	if (iter == m_ConnectionList.end())
+	{
+		// The SYN flag in the first packet must be set.
+		if (! isSyn)
+		{
+			LOG_DEBUG("Ignore packet of a flow [0x%X] that doesn't start with 3 way handshake.", flowKey);
+			return;
+		}
+		// create a TcpSorterData object and add it to the active connection list
+		tcpSorterData = std::make_shared<TcpSorterData>();
```

Comment:
If a `unique_ptr` is acceptiable in this case you should use it instead of `shared_ptr`